data_structures/binary_tree/diff_views_of_binary_tree.py

```diff
@@ -0,0 +1,253 @@
+    ) -> None:
+        self.val = val
+        self.left = left
+        self.right = right
+
+
+def binary_tree_right_side_view(root: TreeNode | None) -> list[int]:
+    """
+    Function returns the right side view of binary tree.
+
+    >>> binary_tree_right_side_view(None)
+    []
+
+    >>> tree_1 = TreeNode(3)
+    >>> tree_1.left = TreeNode(9)
+    >>> tree_1.right = TreeNode(20)
+    >>> tree_1.right.left = TreeNode(15)
+    >>> tree_1.right.right = TreeNode(7)
+    >>> binary_tree_right_side_view(tree_1)
+    [3, 20, 7]
+    """
+
+    def dfs(root: TreeNode | None, depth: int, right_view: list[int]) -> None:
+        """
+        A depth first search preorder traversal to append the values at
+        right side of tree.
+
+        >>> dfs([], 0, [])
+        None
+        """
+        if not root:
+            return
+
+        if depth == len(right_view):
+            right_view.append(root.val)
+
+        dfs(root.right, depth + 1, right_view)
+        dfs(root.left, depth + 1, right_view)
+
+    right_view: list = []
+    if not root:
+        return right_view
+    dfs(root, 0, right_view)
+    return right_view
+
+
+def binary_tree_left_side_view(root: TreeNode | None) -> list[int]:
+    """
+    Function returns the left side view of binary tree.
+
+    >>> binary_tree_left_side_view(None)
+    []
+
+
+    >>> tree_1 = TreeNode(3)
+    >>> tree_1.left = TreeNode(9)
+    >>> tree_1.right = TreeNode(20)
+    >>> tree_1.right.left = TreeNode(15)
+    >>> tree_1.right.right = TreeNode(7)
+    >>> binary_tree_left_side_view(tree_1)
+    [3, 9, 15]
+    """
+
+    def dfs(root: TreeNode | None, depth: int, left_view: list[int]) -> None:
+        """
+        A depth first search preorder traversal to append the values
+        at left side of tree.
+
+        >>> dfs([], 0, [])
+        None
+        """
+        if not root:
+            return
+
+        if depth == len(left_view):
+            left_view.append(root.val)
+
+        dfs(root.left, depth + 1, left_view)
+        dfs(root.right, depth + 1, left_view)
+
+    left_view: list = []
+    if not root:
+        return left_view
+    dfs(root, 0, left_view)
+    return left_view
+
+
+def binary_tree_top_side_view(root: TreeNode | None) -> list[int]:
+    """
+    Function returns the top side view of binary tree.
+
+    >>> binary_tree_top_side_view(None)
+    []
+
+
+    >>> tree_1 = TreeNode(3)
+    >>> tree_1.left = TreeNode(9)
+    >>> tree_1.right = TreeNode(20)
+    >>> tree_1.right.left = TreeNode(15)
+    >>> tree_1.right.right = TreeNode(7)
+    >>> binary_tree_top_side_view(tree_1)
+    [9, 3, 20, 7]
+    """
+    from collections import defaultdict
+
+    def bfs(root: TreeNode, top_view: list[int]) -> None:
+        """
+        A breadth first search traversal with defaultdict ds to append
+        the values of tree from top view
+
+        >>> bfs(TreeNode(5), [])
+        None
```

Comment:
I do not think doctests run on subfunctions.

Reply:
yes, its true it wont run on subfunctions

Reply:
Algorithm-keeper bot asking for same